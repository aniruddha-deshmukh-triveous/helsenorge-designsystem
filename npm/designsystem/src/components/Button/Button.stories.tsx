import React from 'react';

import { ComponentStory, ComponentMeta } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import Button, { ButtonVariants, ButtonIntents } from './Button';
import Icon from '../Icons';
import X from '../Icons/X';
import Calendar from '../Icons/Calendar';

const allButtonVariants: ButtonVariants[] = ['fill', 'outline', 'borderless'];
const allButtonIntents: ButtonIntents[] = ['primary', 'warning', 'danger'];

export default {
  title: 'Button',
  component: Button,
  argTypes: {
    children: {
      control: 'text',
      defaultValue: 'Button',
    },
    ellipsis: {
      control: 'boolean',
      defaultValue: false,
    },
    disabled: {
      control: 'boolean',
      defaultValue: false,
    },
    fluid: {
      control: 'boolean',
      defaultValue: false,
    },
    intent: {
      control: 'select',
      options: allButtonIntents,
      defaultValue: 'primary',
    },
    inverted: {
      control: 'boolean',
      defaultValue: false,
    },
    htmlMarkup: {
      control: 'select',
      options: ['button', 'a'],
      defaultValue: 'button',
    },
    large: {
      control: 'boolean',
      defaultValue: false,
    },
    loading: {
      control: 'boolean',
      defaultValue: false,
    },
    variant: {
      control: 'select',
      options: allButtonVariants,
      defaultValue: 'fill',
    },
    href: {
      control: 'text',
      defaultValue: 'https://www.helsenorge.no',
    },
  },
} as ComponentMeta<typeof Button>;

const myRef1: React.RefObject<HTMLButtonElement> = React.createRef();
const myRef2: React.RefObject<HTMLButtonElement> = React.createRef();
const myRef3: React.RefObject<HTMLButtonElement> = React.createRef();
const myRef4: React.RefObject<HTMLButtonElement> = React.createRef();

export const Default: ComponentStory<typeof Button> = (args: any) => (
  <div style={{ width: '15rem' }}>
    <Button {...args} onClick={action('Button clicked')}>
      {args.children}
    </Button>
  </div>
);

export const Fluid: ComponentStory<typeof Button> = (args: any) => (
  <div style={{ width: '30rem', padding: '1rem' }}>
    <Button {...args} fluid onClick={action('Button clicked')}>
      {args.children}
    </Button>
    <br />
    <br />
    <Button {...args} fluid onClick={action('Button clicked')}>
      <Icon svgIcon={X} />
      {args.children}
      <Icon svgIcon={X} />
    </Button>
    <br />
    <br />
    <Button {...args} fluid onClick={action('Button clicked')}>
      <Icon svgIcon={X} />
      {args.children}
    </Button>
  </div>
);

export const WithIcons: ComponentStory<typeof Button> = (args: any) => (
  <div>
    <Button {...args} ref={myRef1} onClick={action('Button clicked')}>
      <Icon svgIcon={X} />
      {args.children}
    </Button>
    <br />
    <br />
    <Button {...args} ref={myRef2} onClick={action('Button clicked')}>
      {args.children}
      <Icon svgIcon={X} />
    </Button>
    <br />
    <br />
    <Button {...args} ref={myRef3} onClick={action('Button clicked')}>
      <Icon svgIcon={Calendar} />
      {args.children}
      <Icon svgIcon={X} />
    </Button>
    <br />
    <br />
    <Button {...args} ref={myRef4} onClick={action('Button clicked')}>
      <Icon svgIcon={X} />
    </Button>
  </div>
);

export const AllVariants: ComponentStory<typeof Button> = (args: any) => (
  <div style={{ display: 'grid', gridGap: '2rem' }}>
    <Button {...args} variant={'fill'} onClick={action('Button clicked')}>
      <Icon svgIcon={X} />
      {'Fill'}
    </Button>
    <Button {...args} variant={'outline'} onClick={action('Button clicked')}>
      {'Outline'}
      <Icon svgIcon={X} />
    </Button>
    <Button {...args} variant={'borderless'} onClick={action('Button clicked')}>
      <Icon svgIcon={Calendar} />
      {'Borderless'}
      <Icon svgIcon={X} />
    </Button>
  </div>
);
