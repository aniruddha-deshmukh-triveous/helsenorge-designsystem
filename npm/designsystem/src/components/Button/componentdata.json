{
  "props": {
    "children": {
      "defaultValue": null,
      "description": "Sets the content of the button.",
      "name": "children",
      "parent": {
        "fileName": "src/components/Button/Button.tsx",
        "name": "ButtonProps"
      },
      "declarations": [
        {
          "fileName": "src/components/Button/Button.tsx",
          "name": "ButtonProps"
        }
      ],
      "required": true,
      "type": {
        "name": "ReactNode"
      }
    },
    "className": {
      "defaultValue": null,
      "description": "Adds custom classes to the element.",
      "name": "className",
      "parent": {
        "fileName": "src/components/Button/Button.tsx",
        "name": "ButtonProps"
      },
      "declarations": [
        {
          "fileName": "src/components/Button/Button.tsx",
          "name": "ButtonProps"
        }
      ],
      "required": false,
      "type": {
        "name": "string"
      }
    },
    "fluid": {
      "defaultValue": null,
      "description": "Makes the button scale to full width of its parent element.",
      "name": "fluid",
      "parent": {
        "fileName": "src/components/Button/Button.tsx",
        "name": "ButtonProps"
      },
      "declarations": [
        {
          "fileName": "src/components/Button/Button.tsx",
          "name": "ButtonProps"
        }
      ],
      "required": false,
      "type": {
        "name": "boolean"
      }
    },
    "intent": {
      "defaultValue": null,
      "description": "Changes the intent of the button. Mostly changes the color profile.",
      "name": "intent",
      "parent": {
        "fileName": "src/components/Button/Button.tsx",
        "name": "ButtonProps"
      },
      "declarations": [
        {
          "fileName": "src/components/Button/Button.tsx",
          "name": "ButtonProps"
        }
      ],
      "required": false,
      "type": {
        "name": "enum",
        "raw": "\"primary\" | \"warning\" | \"danger\" | undefined",
        "value": [
          {
            "value": "undefined"
          },
          {
            "value": "\"primary\""
          },
          {
            "value": "\"warning\""
          },
          {
            "value": "\"danger\""
          }
        ]
      }
    },
    "inverted": {
      "defaultValue": null,
      "description": "Changes the color presentation for different backgrounds.",
      "name": "inverted",
      "parent": {
        "fileName": "src/components/Button/Button.tsx",
        "name": "ButtonProps"
      },
      "declarations": [
        {
          "fileName": "src/components/Button/Button.tsx",
          "name": "ButtonProps"
        }
      ],
      "required": false,
      "type": {
        "name": "boolean"
      }
    },
    "htmlMarkup": {
      "defaultValue": null,
      "description": "Changes the underlying element of the button.",
      "name": "htmlMarkup",
      "parent": {
        "fileName": "src/components/Button/Button.tsx",
        "name": "ButtonProps"
      },
      "declarations": [
        {
          "fileName": "src/components/Button/Button.tsx",
          "name": "ButtonProps"
        }
      ],
      "required": false,
      "type": {
        "name": "enum",
        "raw": "\"button\" | \"a\" | undefined",
        "value": [
          {
            "value": "undefined"
          },
          {
            "value": "\"button\""
          },
          {
            "value": "\"a\""
          }
        ]
      }
    },
    "large": {
      "defaultValue": null,
      "description": "Changes the size of the button to large.",
      "name": "large",
      "parent": {
        "fileName": "src/components/Button/Button.tsx",
        "name": "ButtonProps"
      },
      "declarations": [
        {
          "fileName": "src/components/Button/Button.tsx",
          "name": "ButtonProps"
        }
      ],
      "required": false,
      "type": {
        "name": "boolean"
      }
    },
    "loading": {
      "defaultValue": null,
      "description": "Sets the button into a loading state displaying <Loader /> as content.",
      "name": "loading",
      "parent": {
        "fileName": "src/components/Button/Button.tsx",
        "name": "ButtonProps"
      },
      "declarations": [
        {
          "fileName": "src/components/Button/Button.tsx",
          "name": "ButtonProps"
        }
      ],
      "required": false,
      "type": {
        "name": "boolean"
      }
    },
    "onClick": {
      "defaultValue": null,
      "description": "",
      "name": "onClick",
      "parent": {
        "fileName": "src/components/Button/Button.tsx",
        "name": "ButtonProps"
      },
      "declarations": [
        {
          "fileName": "src/components/Button/Button.tsx",
          "name": "ButtonProps"
        }
      ],
      "required": false,
      "type": {
        "name": "((e?: MouseEvent<HTMLElement, MouseEvent>) => void) | undefined"
      }
    },
    "variant": {
      "defaultValue": null,
      "description": "Changes the visual representation of the button.",
      "name": "variant",
      "parent": {
        "fileName": "src/components/Button/Button.tsx",
        "name": "ButtonProps"
      },
      "declarations": [
        {
          "fileName": "src/components/Button/Button.tsx",
          "name": "ButtonProps"
        }
      ],
      "required": false,
      "type": {
        "name": "enum",
        "raw": "\"fill\" | \"outline\" | \"borderless\" | undefined",
        "value": [
          {
            "value": "undefined"
          },
          {
            "value": "\"fill\""
          },
          {
            "value": "\"outline\""
          },
          {
            "value": "\"borderless\""
          }
        ]
      }
    },
    "ellipsis": {
      "defaultValue": null,
      "description": "Disables text wrapping and enables ellipsis.",
      "name": "ellipsis",
      "parent": {
        "fileName": "src/components/Button/Button.tsx",
        "name": "ButtonProps"
      },
      "declarations": [
        {
          "fileName": "src/components/Button/Button.tsx",
          "name": "ButtonProps"
        }
      ],
      "required": false,
      "type": {
        "name": "boolean"
      }
    },
    "testId": {
      "defaultValue": null,
      "description": "Sets the data-testid attribute.",
      "name": "testId",
      "parent": {
        "fileName": "src/components/Button/Button.tsx",
        "name": "ButtonProps"
      },
      "declarations": [
        {
          "fileName": "src/components/Button/Button.tsx",
          "name": "ButtonProps"
        }
      ],
      "required": false,
      "type": {
        "name": "string"
      }
    },
    "autoFocus": {
      "defaultValue": null,
      "description": "",
      "name": "autoFocus",
      "parent": {
        "fileName": "designsystem/src/constants.ts",
        "name": "HTMLButtonProps"
      },
      "declarations": [
        {
          "fileName": "designsystem/src/constants.ts",
          "name": "HTMLButtonProps"
        }
      ],
      "required": false,
      "type": {
        "name": "boolean"
      }
    },
    "disabled": {
      "defaultValue": null,
      "description": "",
      "name": "disabled",
      "parent": {
        "fileName": "designsystem/src/constants.ts",
        "name": "HTMLButtonProps"
      },
      "declarations": [
        {
          "fileName": "designsystem/src/constants.ts",
          "name": "HTMLButtonProps"
        }
      ],
      "required": false,
      "type": {
        "name": "boolean"
      }
    },
    "form": {
      "defaultValue": null,
      "description": "",
      "name": "form",
      "parent": {
        "fileName": "designsystem/src/constants.ts",
        "name": "HTMLButtonProps"
      },
      "declarations": [
        {
          "fileName": "designsystem/src/constants.ts",
          "name": "HTMLButtonProps"
        }
      ],
      "required": false,
      "type": {
        "name": "string"
      }
    },
    "formAction": {
      "defaultValue": null,
      "description": "",
      "name": "formAction",
      "parent": {
        "fileName": "designsystem/src/constants.ts",
        "name": "HTMLButtonProps"
      },
      "declarations": [
        {
          "fileName": "designsystem/src/constants.ts",
          "name": "HTMLButtonProps"
        }
      ],
      "required": false,
      "type": {
        "name": "string"
      }
    },
    "formEncType": {
      "defaultValue": null,
      "description": "",
      "name": "formEncType",
      "parent": {
        "fileName": "designsystem/src/constants.ts",
        "name": "HTMLButtonProps"
      },
      "declarations": [
        {
          "fileName": "designsystem/src/constants.ts",
          "name": "HTMLButtonProps"
        }
      ],
      "required": false,
      "type": {
        "name": "string"
      }
    },
    "formMethod": {
      "defaultValue": null,
      "description": "",
      "name": "formMethod",
      "parent": {
        "fileName": "designsystem/src/constants.ts",
        "name": "HTMLButtonProps"
      },
      "declarations": [
        {
          "fileName": "designsystem/src/constants.ts",
          "name": "HTMLButtonProps"
        }
      ],
      "required": false,
      "type": {
        "name": "string"
      }
    },
    "formNoValidate": {
      "defaultValue": null,
      "description": "",
      "name": "formNoValidate",
      "parent": {
        "fileName": "designsystem/src/constants.ts",
        "name": "HTMLButtonProps"
      },
      "declarations": [
        {
          "fileName": "designsystem/src/constants.ts",
          "name": "HTMLButtonProps"
        }
      ],
      "required": false,
      "type": {
        "name": "boolean"
      }
    },
    "formTarget": {
      "defaultValue": null,
      "description": "",
      "name": "formTarget",
      "parent": {
        "fileName": "designsystem/src/constants.ts",
        "name": "HTMLButtonProps"
      },
      "declarations": [
        {
          "fileName": "designsystem/src/constants.ts",
          "name": "HTMLButtonProps"
        }
      ],
      "required": false,
      "type": {
        "name": "string"
      }
    },
    "name": {
      "defaultValue": null,
      "description": "",
      "name": "name",
      "parent": {
        "fileName": "designsystem/src/constants.ts",
        "name": "HTMLButtonProps"
      },
      "declarations": [
        {
          "fileName": "designsystem/src/constants.ts",
          "name": "HTMLButtonProps"
        }
      ],
      "required": false,
      "type": {
        "name": "string"
      }
    },
    "type": {
      "defaultValue": null,
      "description": "",
      "name": "type",
      "parent": {
        "fileName": "designsystem/src/constants.ts",
        "name": "HTMLButtonProps"
      },
      "declarations": [
        {
          "fileName": "designsystem/src/constants.ts",
          "name": "HTMLButtonProps"
        }
      ],
      "required": false,
      "type": {
        "name": "string"
      }
    },
    "value": {
      "defaultValue": null,
      "description": "",
      "name": "value",
      "parent": {
        "fileName": "designsystem/src/constants.ts",
        "name": "HTMLButtonProps"
      },
      "declarations": [
        {
          "fileName": "designsystem/src/constants.ts",
          "name": "HTMLButtonProps"
        }
      ],
      "required": false,
      "type": {
        "name": "string | number | string[]"
      }
    },
    "download": {
      "defaultValue": null,
      "description": "",
      "name": "download",
      "parent": {
        "fileName": "designsystem/src/constants.ts",
        "name": "HTMLAnchorProps"
      },
      "declarations": [
        {
          "fileName": "designsystem/src/constants.ts",
          "name": "HTMLAnchorProps"
        }
      ],
      "required": false,
      "type": {
        "name": "string"
      }
    },
    "href": {
      "defaultValue": null,
      "description": "",
      "name": "href",
      "parent": {
        "fileName": "designsystem/src/constants.ts",
        "name": "HTMLAnchorProps"
      },
      "declarations": [
        {
          "fileName": "designsystem/src/constants.ts",
          "name": "HTMLAnchorProps"
        }
      ],
      "required": false,
      "type": {
        "name": "string"
      }
    },
    "hrefLang": {
      "defaultValue": null,
      "description": "",
      "name": "hrefLang",
      "parent": {
        "fileName": "designsystem/src/constants.ts",
        "name": "HTMLAnchorProps"
      },
      "declarations": [
        {
          "fileName": "designsystem/src/constants.ts",
          "name": "HTMLAnchorProps"
        }
      ],
      "required": false,
      "type": {
        "name": "string"
      }
    },
    "media": {
      "defaultValue": null,
      "description": "",
      "name": "media",
      "parent": {
        "fileName": "designsystem/src/constants.ts",
        "name": "HTMLAnchorProps"
      },
      "declarations": [
        {
          "fileName": "designsystem/src/constants.ts",
          "name": "HTMLAnchorProps"
        }
      ],
      "required": false,
      "type": {
        "name": "string"
      }
    },
    "ping": {
      "defaultValue": null,
      "description": "",
      "name": "ping",
      "parent": {
        "fileName": "designsystem/src/constants.ts",
        "name": "HTMLAnchorProps"
      },
      "declarations": [
        {
          "fileName": "designsystem/src/constants.ts",
          "name": "HTMLAnchorProps"
        }
      ],
      "required": false,
      "type": {
        "name": "string"
      }
    },
    "rel": {
      "defaultValue": null,
      "description": "",
      "name": "rel",
      "parent": {
        "fileName": "designsystem/src/constants.ts",
        "name": "HTMLAnchorProps"
      },
      "declarations": [
        {
          "fileName": "designsystem/src/constants.ts",
          "name": "HTMLAnchorProps"
        }
      ],
      "required": false,
      "type": {
        "name": "string"
      }
    },
    "target": {
      "defaultValue": null,
      "description": "",
      "name": "target",
      "parent": {
        "fileName": "designsystem/src/constants.ts",
        "name": "HTMLAnchorProps"
      },
      "declarations": [
        {
          "fileName": "designsystem/src/constants.ts",
          "name": "HTMLAnchorProps"
        }
      ],
      "required": false,
      "type": {
        "name": "string"
      }
    },
    "referrerPolicy": {
      "defaultValue": null,
      "description": "",
      "name": "referrerPolicy",
      "parent": {
        "fileName": "designsystem/src/constants.ts",
        "name": "HTMLAnchorProps"
      },
      "declarations": [
        {
          "fileName": "designsystem/src/constants.ts",
          "name": "HTMLAnchorProps"
        }
      ],
      "required": false,
      "type": {
        "name": "enum",
        "raw": "\"\" | \"same-origin\" | \"no-referrer\" | \"no-referrer-when-downgrade\" | \"origin\" | \"origin-when-cross-origin\" | \"strict-origin\" | \"strict-origin-when-cross-origin\" | \"unsafe-url\" | undefined",
        "value": [
          {
            "value": "undefined"
          },
          {
            "value": "\"\""
          },
          {
            "value": "\"same-origin\""
          },
          {
            "value": "\"no-referrer\""
          },
          {
            "value": "\"no-referrer-when-downgrade\""
          },
          {
            "value": "\"origin\""
          },
          {
            "value": "\"origin-when-cross-origin\""
          },
          {
            "value": "\"strict-origin\""
          },
          {
            "value": "\"strict-origin-when-cross-origin\""
          },
          {
            "value": "\"unsafe-url\""
          }
        ]
      }
    }
  }
}
