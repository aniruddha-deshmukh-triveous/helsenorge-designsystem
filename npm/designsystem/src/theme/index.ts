import { screen, breakpoints } from './grid';
import { palette } from './palette';
import { spacer, spacers } from './spacers';

export const theme = {
  spacer,
  spacers,
  palette,
  screen,
  breakpoints,
};
