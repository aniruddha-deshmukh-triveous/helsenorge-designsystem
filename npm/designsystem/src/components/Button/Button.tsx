import React from 'react';

import { HTMLButtonProps, HTMLAnchorProps } from '../../constants';
import { getColor } from '../../theme/currys/color';
import { IconProps, IconSize } from './../Icons/';
import Loader from '../Loader';
import { PaletteNames } from '../../theme/palette';
import { useHover } from '../../hooks/useHover';
import { useIcons } from '../../hooks/useIcons';
import { useWindowSize } from '../../hooks/useWindowSize';
import { breakpoints } from '../../theme/grid';
import classNames from 'classnames';

import buttonStyles from './styles.module.scss';

export type ButtonIntents = 'primary' | 'warning' | 'danger';
export type ButtonIntentsColors = 'blueberry' | 'banana' | 'cherry' | 'neutral' | 'white';
export type ButtonTags = 'button' | 'a';
export type ButtonVariants = 'fill' | 'outline' | 'borderless';

interface IntentToColor {
  primary: ButtonIntentsColors;
  warning: ButtonIntentsColors;
  danger: ButtonIntentsColors;
}

export const intentToColor: IntentToColor = {
  primary: 'blueberry',
  warning: 'banana',
  danger: 'cherry',
};

export interface ButtonProps extends HTMLButtonProps, HTMLAnchorProps {
  /** Sets the content of the button. */
  children: React.ReactNode;
  /** Adds custom classes to the element. */
  className?: string;
  /** Makes the button scale to full width of its parent element. */
  fluid?: boolean;
  /** Changes the intent of the button. Mostly changes the color profile. */
  intent?: ButtonIntents;
  /** Changes the color presentation for different backgrounds. */
  inverted?: boolean;
  /** Changes the underlying element of the button. */
  htmlMarkup?: ButtonTags;
  /** Changes the size of the button to large. */
  large?: boolean;
  /** Sets the button into a loading state displaying <Loader /> as content. */
  loading?: boolean;
  onClick?: (e?: React.MouseEvent<HTMLElement, MouseEvent>) => void;
  /** Changes the visual representation of the button. */
  variant?: ButtonVariants;
  /** Disables text wrapping and enables ellipsis. */
  ellipsis?: boolean;
  /** Sets the data-testid attribute. */
  testId?: string;
}

const getIconColor = (fill: boolean, disabled: boolean, intent: ButtonIntents, inverted: boolean, hovered: boolean): string => {
  if (disabled) return getColor('neutral', 600);
  if ((fill && !inverted) || (!fill && inverted)) return 'white';
  return getColor(intentToColor[intent], hovered ? 700 : 600);
};

const getLargeIconSize = (large: boolean, screenWidth: number | undefined): number => {
  const mobile = screenWidth && screenWidth < breakpoints.md;
  if (mobile && large) return IconSize.Small;
  if (large) return IconSize.Medium;
  return IconSize.XSmall;
};

const Button = React.forwardRef(function ButtonForwardedRef(
  props: ButtonProps,
  ref: React.ForwardedRef<HTMLButtonElement | HTMLAnchorElement>
) {
  const {
    children,
    className = '',
    fluid = false,
    intent = 'primary',
    inverted = false,
    htmlMarkup = 'button',
    onClick,
    large = false,
    loading = false,
    variant = 'fill',
    disabled = false,
    ellipsis = false,
    testId,
    href,
    target,
    ...restProps
  } = props;

  const [leftIcon, rightIcon, restChildren] = useIcons(React.Children.toArray(children));
  const { hoverRef, isHovered } =
    htmlMarkup === 'button'
      ? useHover<HTMLButtonElement>(ref as React.RefObject<HTMLButtonElement>)
      : useHover<HTMLAnchorElement>(ref as React.RefObject<HTMLAnchorElement>);
  const iconColor = getIconColor(variant === 'fill', disabled, intent, inverted, isHovered);
  const size = useWindowSize();
  const fillVariant = variant === 'fill';
  const outlineVariant = variant === 'outline';
  const borderlessVariant = variant === 'borderless';
  const warningIntent = intent === 'warning';
  const dangerIntent = intent === 'danger';
  const hasIcon = !!(leftIcon || rightIcon) && !loading;
  const rest = { ...(restProps as React.HtmlHTMLAttributes<HTMLButtonElement>) };

  const buttonClasses = classNames(
    buttonStyles.button,
    {
      [buttonStyles['button--fluid']]: fluid,
      [buttonStyles['button--large']]: large,
      [buttonStyles[`button--warning`]]: warningIntent,
      [buttonStyles[`button--danger`]]: dangerIntent,
      [buttonStyles[`button--inverted`]]: inverted,
      [buttonStyles['button--fill']]: fillVariant,
      [buttonStyles['button--outline']]: outlineVariant,
      [buttonStyles['button--borderless']]: borderlessVariant,
      [buttonStyles['button--with-icon']]: hasIcon,
      [buttonStyles['button--ellipsis']]: ellipsis,
    },
    className
  );

  const contentClasses = classNames(buttonStyles.button__content, {
    [buttonStyles['button__content--warning']]: warningIntent,
    [buttonStyles['button__content--danger']]: dangerIntent,
    [buttonStyles['button__content--fill']]: fillVariant,
    [buttonStyles['button__content--borderless']]: borderlessVariant,
    [buttonStyles['button__content--with-icon']]: hasIcon,
    [buttonStyles['button__content--large']]: large,
    [buttonStyles['button__content--ellipsis']]: ellipsis,
    [buttonStyles['button__content--inverted']]: inverted,
    [buttonStyles['button__content--fluid']]: fluid,
  });

  const leftFluidContentClasses = classNames(buttonStyles['button__left-fluid-content'], {
    [buttonStyles['button__left-fluid-content--with-icon']]: hasIcon,
    [buttonStyles['button__left-fluid-content--large']]: large,
  });

  function renderIcon(
    iconElement: React.ReactElement<IconProps> | {} | undefined | null,
    size: number,
    iconColor: string,
    hover: boolean
  ): React.ReactElement<IconProps> | React.Component<IconProps> | null {
    const color =
      iconElement && (iconElement as React.ReactElement<IconProps>).props && (iconElement as React.ReactElement<IconProps>).props.color
        ? (iconElement as React.ReactElement<IconProps>).props.color
        : iconColor;
    return iconElement && Object.keys(iconElement).length > 0
      ? React.cloneElement(iconElement as React.ReactElement<IconProps>, { size, color, isHovered: hover })
      : null;
  }

  const renderButtonContent = () => {
    return restChildren ? <span className={contentClasses}>{restChildren}</span> : null;
  };

  const renderbuttonContentWrapper = () => {
    return (
      <span className={buttonStyles['content-wrapper']}>
        {loading ? (
          <div className={buttonStyles['button__left-fluid-content']}>
            <Loader
              testId={'test-id-loader'}
              color={variant === 'fill' && !inverted ? 'white' : (intentToColor[intent] as PaletteNames)}
              size="tiny"
            />
          </div>
        ) : (
          <>
            {fluid ? (
              <div className={leftFluidContentClasses}>
                {renderIcon(leftIcon, getLargeIconSize(large, size.width), iconColor, isHovered)}
                {renderButtonContent()}
              </div>
            ) : (
              <>
                {renderIcon(leftIcon, getLargeIconSize(large, size.width), iconColor, isHovered)}
                {renderButtonContent()}
              </>
            )}
            {renderIcon(rightIcon, IconSize.XSmall, iconColor, isHovered)}
          </>
        )}
      </span>
    );
  };

  return (
    <>
      {htmlMarkup === 'button' && (
        <button
          onClick={onClick}
          disabled={disabled}
          data-testid={testId}
          className={buttonClasses}
          ref={hoverRef as React.ForwardedRef<HTMLButtonElement>}
          {...rest}
        >
          {renderbuttonContentWrapper()}
        </button>
      )}
      {htmlMarkup === 'a' && (
        <a
          onClick={onClick}
          data-testid={testId}
          className={buttonClasses}
          href={href}
          target={target}
          rel={target === '_blank' ? 'noopener noreferrer' : props.rel}
          ref={hoverRef as React.ForwardedRef<HTMLAnchorElement>}
          {...restProps}
        >
          {renderbuttonContentWrapper()}
        </a>
      )}
    </>
  );
});

export default Button;
