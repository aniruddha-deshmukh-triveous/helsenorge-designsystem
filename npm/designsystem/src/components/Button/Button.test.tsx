import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import Button from './Button';
import Icon from '../Icons';
import Check from '../Icons/Check';
import { palette } from '../../theme/palette';

describe('Gitt at button skal vises', (): void => {
  describe('Når button rendres', (): void => {
    test('Så vises button', (): void => {
      const { container } = render(<Button>Button</Button>);
      expect(container).toMatchSnapshot();
    });

    test('Så er textwrap på', (): void => {
      render(<Button>Button text</Button>);

      const testButtonText = screen.getByText('Button text');
      expect(testButtonText.className).toBe('button__content button__content--fill button__content--fill-not-inverted');
    });
  });

  describe('Når button rendres med ikoner', (): void => {
    test('Så vises ikonene', (): void => {
      const { container } = render(
        <Button>
          <Icon svgIcon={Check} />
          Button
          <Icon svgIcon={Check} />
        </Button>
      );
      expect(container).toMatchSnapshot();
    });
  });

  describe('Når button rendres med ellipsis på', (): void => {
    test('Så brukes ellipsis på overflødig tekst', (): void => {
      render(<Button ellipsis={true}>Button text</Button>);

      const testButtonText = screen.getByText('Button text');
      expect(testButtonText.className).toBe('button__content button__content--fill button__content--fill-not-inverted');
    });
  });

  describe('Når button rendres med fluid og ikon(er)', (): void => {
    test('Så venstrestilles teksten', (): void => {
      render(
        <Button testId={'test01'} fluid={true}>
          Button text
          <Icon svgIcon={Check} />
        </Button>
      );

      const testLeftFluidContent = screen.getByTestId('test01').querySelector('div');
      expect(testLeftFluidContent.className).toBe('button__left-fluid-content button__left-fluid-content--with-icon');
    });
  });

  describe('Når button rendres med disabled', (): void => {
    test('Så kan ikke knappen trykkes på', (): void => {
      const handleClick = jest.fn();

      render(
        <Button testId={'test01'} onClick={handleClick} disabled={true}>
          Button
        </Button>
      );

      const testButton = screen.getByTestId('test01');

      fireEvent.click(testButton);

      expect(handleClick).toHaveBeenCalledTimes(0);
      expect(testButton).toBeDisabled();
    });
  });

  describe('Når button rendres med intent', (): void => {
    test('Så er background-color riktig', (): void => {
      render(
        <div>
          <Button testId={'test01'} intent={'primary'}>
            Button
          </Button>
          <Button testId={'test02'} intent={'warning'}>
            Button
          </Button>
          <Button testId={'test03'} intent={'danger'}>
            Button
          </Button>
        </div>
      );

      const testButton1 = screen.getByTestId('test01');
      const testButton2 = screen.getByTestId('test02');
      const testButton3 = screen.getByTestId('test03');

      expect(testButton1.className).toBe('button button--fill button--fill-not-inverted');
      expect(testButton2.className).toBe('button button--fill button--fill-warning button--fill-not-inverted');
      expect(testButton3.className).toBe('button button--fill button--fill-danger button--fill-not-inverted');
    });
  });

  describe('Når button rendres med inverted', (): void => {
    test('Så er farger riktig', (): void => {
      render(
        <Button testId={'test01'} inverted={true}>
          Button
        </Button>
      );

      const testButton = screen.getByTestId('test01');

      expect(testButton.className).toBe('button button--fill button--fill-inverted');
    });
  });

  describe('Når button rendres med large', (): void => {
    test('Så er ikon riktig størrelse', (): void => {
      render(
        <Button large={true}>
          <Icon testId={'test01'} svgIcon={Check} />
          Button
        </Button>
      );

      const testButtonIcon = screen.getByTestId('test01');
      const height = getComputedStyle(testButtonIcon).getPropertyValue('min-height');
      const width = getComputedStyle(testButtonIcon).getPropertyValue('min-width');

      expect(height).toBe('64px');
      expect(width).toBe('64px');
    });
  });

  describe('Når button rendres med loading', (): void => {
    test('Så er loader elementet lagt til', (): void => {
      render(
        <Button loading={true}>
          <Icon svgIcon={Check} />
          Button
        </Button>
      );

      const testLoader = screen.getByTestId('test-id-loader');

      expect(testLoader).toBeTruthy();
    });
  });

  describe('Når button rendres med varianter', (): void => {
    test('Så rendres innholdet riktig', (): void => {
      render(
        <div>
          <Button testId={'test01'} variant={'fill'}>
            Button
          </Button>
          <Button testId={'test02'} variant={'outline'}>
            Button
          </Button>
          <Button testId={'test03'} variant={'borderless'}>
            Button
          </Button>
        </div>
      );

      const testFill = screen.getByTestId('test01');

      expect(testFill.className).toBe('button button--fill button--fill-not-inverted');

      const testOutline = screen.getByTestId('test02');

      expect(testOutline.className).toBe('button button--outline');

      const testBorderless = screen.getByTestId('test03');

      expect(testBorderless.className).toBe('button button--borderless');
    });
  });
});
