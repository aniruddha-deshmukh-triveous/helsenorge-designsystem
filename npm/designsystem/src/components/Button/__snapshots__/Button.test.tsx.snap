// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gitt at button skal vises N책r button rendres S책 vises button 1`] = `
<div>
  <button
    class="button button--fill button--fill-not-inverted"
  >
    <span
      class="content-wrapper"
    >
      <span
        class="button__content button__content--fill button__content--fill-not-inverted"
      >
        Button
      </span>
    </span>
  </button>
</div>
`;

exports[`Gitt at button skal vises N책r button rendres med ikoner S책 vises ikonene 1`] = `
<div>
  <button
    class="button button--fill button--fill-not-inverted button--with-icon"
  >
    <span
      class="content-wrapper"
    >
      <svg
        aria-hidden="true"
        class="sc-AxiKw kgEvoS hnds-style-icon "
        fill="white"
        height="38"
        role="img"
        viewBox="0 0 48 48"
        width="38"
      >
        <path
          d="M22.504 31.578l-9.781-10.162 1.822-1.752 7.959 8.269 14.265-14.819 1.821 1.752z"
        />
      </svg>
      <span
        class="button__content button__content--fill button__content--fill-not-inverted button__content--with-icon"
      >
        Button
      </span>
      <svg
        aria-hidden="true"
        class="sc-AxiKw kgEvoS hnds-style-icon "
        fill="white"
        height="38"
        role="img"
        viewBox="0 0 48 48"
        width="38"
      >
        <path
          d="M22.504 31.578l-9.781-10.162 1.822-1.752 7.959 8.269 14.265-14.819 1.821 1.752z"
        />
      </svg>
    </span>
  </button>
</div>
`;
